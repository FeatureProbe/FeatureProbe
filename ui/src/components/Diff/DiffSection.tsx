import { cloneDeep } from 'lodash';
import { ReactNode, useEffect, useMemo, useState } from 'react';
import idiff, { ChangeItem } from './diff';
import styles from './DiffSection.module.scss';

export interface DiffSectionProps {
  title: string;
  before?: unknown;
  after?: unknown;
  diffKey: string;
  renderContent: (diffContent: ChangeItem[]) => ReactNode;
  beforeDiff?: (before: unknown, after: unknown) => unknown[];
  setCount?: (key: string, count: number) => void;
}

const DiffSection: React.FC<DiffSectionProps> = (props) => {
  const { title, before, after, renderContent, beforeDiff, setCount, diffKey } = props;
  const [hide, setHide] = useState(false);

  const diffContent = useMemo(() => {
    let left, right;
    if (beforeDiff) {
      [left, right] = beforeDiff(cloneDeep(before), cloneDeep(after));
    } else {
      [left, right] = [before, after];
    }

    const diffContent = before && after ? idiff(left, right) : undefined;

    return diffContent;
  }, [beforeDiff, before, after]);

  useEffect(() => {
    let modifyCount = 0;
    diffContent?.forEach((item) => {
      if (item.type !== 'same') {
        if (diffKey === 'status' || diffKey === 'default' || diffKey === 'disabled' || diffKey === 'prerequisites') {
          modifyCount = 1;
        } else {
          modifyCount++;
        }
      }
    });

    setHide(modifyCount === 0);
    setCount && setCount(diffKey, modifyCount);
  }, [diffContent, diffKey, setCount]);

  return (
    <div hidden={hide} className={styles['diff-section']}>
      <div className={styles['head-line']}>
        <div className={styles['before']}>
          <div className={styles.title}>{title}</div>
        </div>
        <div className={styles['after']}>
          <div className={styles.title}>{title}</div>
        </div>
      </div>
      <div className={styles.content}>{diffContent && renderContent(diffContent)}</div>
    </div>
  );
};

export default DiffSection;
