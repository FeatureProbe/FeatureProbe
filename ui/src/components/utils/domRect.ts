export class DOMRect {
    x: number;
    width: number;
    y: number;
    height: number;
    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}