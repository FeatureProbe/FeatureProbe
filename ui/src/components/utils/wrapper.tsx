import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
