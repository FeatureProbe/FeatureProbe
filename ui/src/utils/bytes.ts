export function getBytes(str: string) {
  let bytes = str.length,
    i = 0;
  for (; i < bytes; i++) if (str.charCodeAt(i) > 255) bytes++;

  return bytes;
}

const MB = 1048576;
export const MAX_SIZE = MB;