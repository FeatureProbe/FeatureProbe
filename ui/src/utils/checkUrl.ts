function escapeStringRegexp(string: string): string {
  if (typeof string !== 'string') {
    throw new TypeError('Expected a string');
  }

  return string
    .replace(/[|\\{}()[\]^$+*?.]/g, '\\$&')
    .replace(/-/g, '\\x2d');
}

export function matchUrl(matcher: string, originUrl: string, targetUrl: string): boolean {
  let href = '';
  let hash = '';
  let search = '';

  try {
    const fullUrl = new URL(targetUrl, '');
    href = fullUrl.href;
    hash = fullUrl.hash;
    search = fullUrl.search;
  } catch (error) {
    href = targetUrl;
  }

  let regex;
  let testUrl;

  switch (matcher) {
    case 'EXACT':
      testUrl = href;
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SIMPLE':
      testUrl = href.replace(hash, '').replace(search, '');
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SUBSTRING':
      testUrl = href.replace(search, '');
      regex = new RegExp('.*' + escapeStringRegexp(originUrl) + '.*$');
      break;
    case 'REGULAR':
      testUrl = href.replace(search, '');
      regex = new RegExp(originUrl);
      break;
    default:
      return false;
  }

  return regex.test(testUrl);
}