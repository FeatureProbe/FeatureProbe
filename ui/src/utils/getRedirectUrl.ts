import { getProjectList } from 'services/project';
import { IProject, IEnvironment } from 'interfaces/project';
import { IDictionary } from 'interfaces/targeting';
import { getFromDictionary } from 'services/dictionary';
import { LAST_SEEN } from 'constants/dictionary_keys';

const getInfo = async () => {
  const res =  await getFromDictionary<IDictionary>(LAST_SEEN);
  const { success, data } = res;
  if (success && data) {
    return JSON.parse(data.value);
  } else {
    return {};
  }
};

export const getRedirectUrl = async (defaultRedirectUrl: string) => {
  let redirectUrl = defaultRedirectUrl;
  const info = await getInfo();
  const { projectKey, environmentKey } = info;

  if (!!projectKey && !!environmentKey) {
    redirectUrl = `/${projectKey}/${environmentKey}/toggles`;
  } else {
    const res = await getProjectList<IProject[]>();
    const { success, data } = res;
    if (success && data) {
      if (data[0]) {
        const { key, environments } = data[0];
        let envKey = '';
        environments.some((environment: IEnvironment) => {
          if (environment.key) {
            envKey = environment.key;
            return true;
          }
          return environment;
        });
  
        if (key && envKey) {
          redirectUrl = `/${key}/${envKey}/toggles`;
        }
      } else {
        redirectUrl = '/projects';
      }
    }
  }

  return redirectUrl;
};