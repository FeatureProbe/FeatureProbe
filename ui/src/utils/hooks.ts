
import { useState } from 'react';

export const useLocalStorage = (key: string, initialValue: boolean | string) => {
  const [storedValue, setStoredValue] = useState(() => {
    if (typeof window === 'undefined') {
      return initialValue;
    }

    try {
      const item = window.localStorage.getItem(key);
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      console.log(error);
      return initialValue;
    }
  });

  const setValue = (value: object) => {
    try {
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);

      if (typeof window !== 'undefined') {
        window.localStorage.setItem(key, JSON.stringify(valueToStore));
      }
      
    } catch (error) {
      console.log(error);
    }
  };
  return [storedValue, setValue];
};
