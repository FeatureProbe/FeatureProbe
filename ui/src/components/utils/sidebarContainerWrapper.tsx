import { SidebarContainer } from 'layout/hooks';

export default SidebarContainer.Provider;
