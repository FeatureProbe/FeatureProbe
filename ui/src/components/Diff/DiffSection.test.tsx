import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import DiffSection from './DiffSection';
import { IntlWrapper } from 'components/utils/wrapper';
import { ReactNode } from 'react';

it('DiffSection snapshot', (done) => {
  (async () => {
    const { asFragment, rerender } = render(
      <DiffSection
        title={'Test'}
        before={['a']}
        after={['a', 'b']}
        diffKey={'Test'}
        renderContent={function (): ReactNode {
          return <div>test</div>;
        }}
      />,
      {
        wrapper: IntlWrapper,
      }
    );
    expect(asFragment()).toMatchSnapshot();

    rerender(
      <DiffSection
        title={'Test'}
        before={['a']}
        after={['a']}
        beforeDiff={() => {
            return [['a'], ['a']];
        }}
        diffKey={'status'}
        renderContent={function (): ReactNode {
          return <div>test</div>;
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});
