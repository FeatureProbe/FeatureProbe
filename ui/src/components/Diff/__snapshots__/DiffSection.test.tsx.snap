// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffSection snapshot 1`] = `
<DocumentFragment>
  <div
    class="diff-section"
  >
    <div
      class="head-line"
    >
      <div
        class="before"
      >
        <div
          class="title"
        >
          Test
        </div>
      </div>
      <div
        class="after"
      >
        <div
          class="title"
        >
          Test
        </div>
      </div>
    </div>
    <div
      class="content"
    >
      <div>
        test
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffSection snapshot 2`] = `
<DocumentFragment>
  <div
    class="diff-section"
    hidden=""
  >
    <div
      class="head-line"
    >
      <div
        class="before"
      >
        <div
          class="title"
        >
          Test
        </div>
      </div>
      <div
        class="after"
      >
        <div
          class="title"
        >
          Test
        </div>
      </div>
    </div>
    <div
      class="content"
    >
      <div>
        test
      </div>
    </div>
  </div>
</DocumentFragment>
`;
